### PR TITLE
MADS: In-memory uncompressor for Rex Nebular

### PR DESCRIPTION
**Pull request by phcoder**

### CODE REVIEW

File: engines/mads/libblast/blast.c [hunk 1]
```diff
@@ -0,0 +1,423 @@
+/* blast.c
+ * Copyright (C) 2003, 2012, 2013 Mark Adler
```
Comment (bluegr) [inline engines/mads/libblast/blast.c:2]: .c? Why?

**Comment by lephilousophe** [comment]
What's the status of this PR rel to #4404?
In addition setjmp.h is forbidden in ScummVM.

File: engines/mads/libblast/blast.c [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+ * 1.1  16 Feb 2003     - Fixed distance check for > 4 GB uncompressed data
+ * 1.2  24 Oct 2012     - Add note about using binary mode in stdio
+ *                      - Fix comparisons of differently signed integers
+ * 1.3  24 Aug 2013     - Return unused input from blast()
+ *                      - Fix test code to correctly report unused input
+ *                      - Enable the provision of initial input to blast()
+ */
+
+#include <stddef.h>             /* for NULL */
+#include <setjmp.h>             /* for setjmp(), longjmp(), and jmp_buf */
```
Comment (bluegr) [inline engines/mads/libblast/blast.c:33]: As mentioned, we avoid using setjmp in ScummVM

File: engines/mads/libblast/blast.c [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+static int bits(struct state *s, int need)
+{
+    int val;            /* bit accumulator */
+
+    /* load at least need bits into val */
+    val = s->bitbuf;
+    while (s->bitcnt < need) {
+        if (s->left == 0) {
+            s->left = s->infun(s->inhow, &(s->in));
+            if (s->left == 0) longjmp(s->env, 1);       /* out of input */
```
Comment (bluegr) [inline engines/mads/libblast/blast.c:80]: We could use `error()` here (since this would indicate corrupt data files), or return -1 and handle this case in the caller

File: engines/mads/libblast/blast.c [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+            first += count;
+            first <<= 1;
+            code <<= 1;
+            len++;
+        }
+        left = (MAXBITS+1) - len;
+        if (left == 0) break;
+        if (s->left == 0) {
+            s->left = s->infun(s->inhow, &(s->in));
+            if (s->left == 0) longjmp(s->env, 1);       /* out of input */
```
Comment (bluegr) [inline engines/mads/libblast/blast.c:164]: Same as above

File: engines/mads/libblast/blast.c (hunk 1 above)
Comment (phcoder) [reply engines/mads/libblast/blast.c]: Moot. The file is removed.

File: engines/mads/libblast/blast.c (hunk 2 above)
Comment (phcoder) [reply engines/mads/libblast/blast.c]: Moot. The file is removed.

File: engines/mads/libblast/blast.c (hunk 4 above)
Comment (phcoder) [reply engines/mads/libblast/blast.c]: Moot. The file is removed.

File: engines/mads/libblast/blast.c (hunk 3 above)
Comment (phcoder) [reply engines/mads/libblast/blast.c]: Moot. The file is removed.

**Comment by phcoder** [comment]
> What's the status of this PR rel to #4404? 
You linked back to this PR
> In addition setjmp.h is forbidden in ScummVM.
Removed

**Comment by phcoder** [comment]
Thank you for pointing to common/dcl.cpp. Yes, it's the same algorithm.

File: engines/mads/detection_tables.h [hunk 5]
```diff
@@ -22,14 +22,13 @@
 namespace MADS {
 
 static const MADSGameDescription gameDescriptions[] = {
-#if 0
 	{
 		// Rex Nebular and the Cosmic Gender Bender DOS English (compressed)
 		// Removed for now, until the original floppy compression is supported
```
Comment (bluegr) [inline engines/mads/detection_tables.h:27]: This can now be removed

```suggestion

```

File: engines/mads/mps_installer.cpp [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+const Common::ArchiveMemberPtr MpsInstaller::getMember(const Common::Path &path) const {
+	Common::String translated = translateName(path);
+	if (!_files.contains(translated))
+		return nullptr;
+
+	return Common::SharedPtr<Common::ArchiveMember>(new Common::GenericArchiveMember(_files.getVal(translated)._fileName, this));
+}
+
+// TODO: Make streams stay valid after destructing of archive
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:88]:
```suggestion
// TODO: Make streams stay valid after destruction of archive
```

File: engines/mads/mads.cpp [hunk 7]
```diff
@@ -77,6 +78,12 @@ MADSEngine::~MADSEngine() {
 }
 
 void MADSEngine::initialize() {
+	if (scumm_stricmp(_gameDescription->desc.filesDescriptions[0].fileName, "mpslabs.001") == 0) {
```
Comment (bluegr) [inline engines/mads/mads.cpp:81]: It would be cleaner to add a game flag for this in the detection tables, e.g. GF_COMPRESSED, instead of checking for data file names

Comment (phcoder) [reply engines/mads/mads.cpp:81]: Done

**Comment by lephilousophe** [comment]
Sorry for the mix up I was referring to scummvm/scummvm-tools#51 and didn't realize that they were in different codebases.

File: common/dcl.h [hunk 8]
```diff
@@ -37,6 +37,7 @@ namespace Common {
  *          - Mohawk
  *          - Neverhood
  *          - SCI
+ *          - MADS (exclusively for REX Nebular installer)
```
Comment (bluegr) [inline common/dcl.h:40]:
```suggestion
 *          - MADS (exclusively for Rex Nebular installer) 
```

File: engines/mads/mps_installer.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	uint32 off = desc._offsetInVolume;
+	uint vol = desc._volumeNumber;
+	uint32 rem = desc._compressedSize;
+	byte *compressedBuf = new byte[rem];
+	byte *outptr = compressedBuf;
+	while (rem > 0) {
+		Common::File fvol;
+		Common::Path volumePath = _baseName.append(Common::String::format(".%03d", vol));
+		if (!fvol.open(volumePath)) {
+			debug("Failed to open volume %s.%03d", volumePath.toString().c_str(), vol);
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:113]: Shouldn't that be an error?

File: engines/mads/mps_installer.cpp [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+			delete[] compressedBuf;
+			return nullptr;
+		}
+		fvol.seek(off);
+		int32 actual = fvol.read(outptr, rem);
+		if (actual <= 0) {
+			debug("Read failure in volume %s.%03d", volumePath.toString().c_str(), vol);
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:120]: Shouldn't that be a warning? We can get more meaningful bug reports if it is shown as a warning in the console

File: engines/mads/mps_installer.cpp [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+		uncompressedSize = desc._compressedSize;
+		compressedBuf = nullptr;
+		break;
+	case 1:
+		Common::MemoryReadStream compressedReadStream(compressedBuf, desc._compressedSize);
+		uncompressedBuf = new byte[uncompressedSize];
+		if (!Common::decompressDCL(&compressedReadStream, uncompressedBuf, desc._compressedSize, uncompressedSize)) {
+			delete[] compressedBuf;
+			delete[] uncompressedBuf;
+			debug ("Unable to decompress %s", desc._fileName.c_str());
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:147]: Shouldn't that be an error?

File: engines/mads/mps_installer.h [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+							       _volumeNumber(FROM_LE_16(raw.volumeNumber)) {}
+		
+		Common::String _fileName;
+		uint _compressionAlgo;
+		uint _volumeNumber;
+		uint32 _offsetInVolume;
+		uint32 _compressedSize;
+		uint32 _uncompressedSize;
+
+		friend class MpsInstaller;
```
Comment (bluegr) [inline engines/mads/mps_installer.h:80]: Why is this needed?

File: engines/mads/mps_installer.cpp (hunk 9 above)
Comment (phcoder) [reply engines/mads/mps_installer.cpp:113]: Done

File: engines/mads/mps_installer.cpp (hunk 10 above)
Comment (phcoder) [reply engines/mads/mps_installer.cpp:120]: Done

File: engines/mads/mps_installer.cpp (hunk 11 above)
Comment (phcoder) [reply engines/mads/mps_installer.cpp:147]: Done

File: engines/mads/mps_installer.h (hunk 12 above)
Comment (phcoder) [reply engines/mads/mps_installer.h:80]: To keep FileDescriptor mostly opaque but allow it to be accessed from MpsInstaller

Comment (bluegr) [reply engines/mads/mps_installer.h:80]: You can access an embedded class from the class it's enclosed in - you don't need a friend class definition for this case. Check for example `video/video_decoder.h`, classes VideoDecoder and Track

Comment (phcoder) [reply engines/mads/mps_installer.h:80]: Without this line I get:
engines/mads/mps_installer.cpp: In static member function ‘static MADS::MpsInstaller* MADS::MpsInstaller::open(const Common::Path&)’:
engines/mads/mps_installer.cpp:52:30: error: ‘MADS::MpsInstaller::FileDescriptor::FileDescriptor(const MADS::FileDescriptorBin&)’ is protected within this context
   52 |   FileDescriptor desc(descBin);

File: engines/mads/mps_installer.cpp [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+	return members;
+}
+
+const Common::ArchiveMemberPtr MpsInstaller::getMember(const Common::Path &path) const {
+	Common::String translated = translateName(path);
+	if (!_files.contains(translated))
+		return nullptr;
+
+	return Common::SharedPtr<Common::ArchiveMember>(new Common::GenericArchiveMember(_files.getVal(translated)._fileName, this));
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:85]:
```suggestion
	return Common::SharedPtr<Common::ArchiveMember>(new Common::GenericArchiveMember(_files.getVal(translated)._fileName, this));
```
    return Common::ArchiveMemberPtr(new Common::GenericArchiveMember(name, this));

Refer to https://github.com/scummvm/scummvm/blob/253a0519a7d403980d265739741155f3f27ba00d/common/installshieldv3_archive.cpp#L137

**Comment by bluegr** [comment]
@phcoder: the last commit references SAGA instead of MADS

**Comment by phcoder** [comment]
> @phcoder: the last commit references SAGA instead of MADS

Done

File: engines/mads/mps_installer.cpp [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+#include "common/util.h"
+#include "common/memstream.h"
+#include "common/dcl.h"
+#include "mps_installer.h"
+
+namespace MADS {
+
+MpsInstaller* MpsInstaller::open(const Common::Path& baseName) {
+	Common::File indexFile;
+	Common::HashMap<Common::String, FileDescriptor, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo> files;
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:35]: Use a typedef, and prefix the variable with _, as per our coding guidelines

```suggestion
	typedef Common::HashMap<Common::String, FileDescriptor, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo> FileMap;
	
	FileMap _files;
```

Check https://github.com/scummvm/scummvm/blob/253a0519a7d403980d265739741155f3f27ba00d/common/installshieldv3_archive.h#L59

File: engines/mads/mps_installer.cpp [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+	uint32 indexSize = indexFile.size();
+
+	if (filecnt > (indexSize - 12) / sizeof(FileDescriptorBin))
+		filecnt = (indexSize - 12) / sizeof(FileDescriptorBin);
+
+	for (uint i = 0; i < filecnt; i++) {
+		FileDescriptorBin descBin;
+		indexFile.read(&descBin, sizeof(descBin));
+		FileDescriptor desc(descBin);
+		files[desc._fileName] = desc;
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:53]:
```suggestion
		_files[desc._fileName] = desc;
```

File: engines/mads/mps_installer.cpp [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	return new MpsInstaller(files, baseName);
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:56]:
```suggestion
	return new MpsInstaller(_files, baseName);
```

File: engines/mads/mps_installer.cpp [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+int MpsInstaller::listMembers(Common::ArchiveMemberList &list) const {
+	int members = 0;
+
+	for (Common::HashMap<Common::String, FileDescriptor, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo>::const_iterator i = _files.begin(), end = _files.end();
+	     i != end; ++i) {
+		list.push_back(Common::ArchiveMemberList::value_type(new Common::GenericArchiveMember(i->_key, this)));
+		++members;
+	}
+
+	return members;
```
Comment (bluegr) [inline engines/mads/mps_installer.cpp:77]: This can be simplified as follows:

	for (FileMap::const_iterator it = _files.begin(); it != _files.end(); it++)
		list.push_back(getMember(it->_key));

	return _map.size();

Check https://github.com/scummvm/scummvm/blob/253a0519a7d403980d265739741155f3f27ba00d/common/installshieldv3_archive.cpp#L132

File: engines/mads/mps_installer.h [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+#include "common/pack-start.h"  // START STRUCT PACKING
+
+struct FileDescriptorBin {
+	char name[0x52]; // zero-terminated, rest is filled with what looks like garbage
+	uint16 compression;
+	uint16 volumeNumber;
+	uint32 offsetInVolume;
+	uint32 compressedSize;
+	uint32 uncompressedSize;
+} PACKED_STRUCT;
```
Comment (bluegr) [inline engines/mads/mps_installer.h:42]: Why do we need this? It would be easier to just have FileDescriptor, and read into it with native endianess. This would simplify the code considerably, and we can remove FileDescriptorBin altogether

**Comment by bluegr** [comment]
Many thanks for your work on this! This can now be merged
